### ISSUE TITLE
openPrimeR

### ISSUE THREAD

**Issue by matdoering** [Bioconductor/Contributions#456]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/matdoering/openPrimeR/

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by lawremi** [comment]
Looks like there are a bunch of packages listed in Imports but not mentioned in NAMESPACE. Probably should just `import(methods)` instead of worrying about individual symbols. Lots of system dependencies. That's going to make the package tough to get working. Should at least consider separating the Shiny app into its own package.

**Comment by matdoering** [comment]
Thanks for having a look! Your suggestion of making the Shiny app an individual package makes a lot of sense. Are there guidelines on how such "sister packages" should be provided? In my case I think it would be enough to ensure that the frontend package requires the correct version of the backend library but I'm wondering if there are other things to consider when doing a split.

Right now, I've just fixed the minor issues (data_raw folder, inst/doc removed) and imported 'methods'. Now, I'm wondering how to debug the OS-specific issues: if I do a version bump for every debug attempt that could be quite a lot of bumps to find out what the problem is. Is there some recommended way to debug OS-specificially if I don't have the OS available (e.g. OSX)? It's interesting that even for Linux the build doesn't work although it runs fine on my system (running R 3.4.0 and not 3.4.1, however).

**Comment by matdoering** [comment]
I've just borrowed a Mac to identify the error that occurs during vignette building. In fact the error is thrown because `GenomeInfoDbData` does not seem to be available when doing a standard installation on a Mac using `devtools::install()`. After manually installing the package using `biocLite()` the vignette building worked fine. The problem is related to lazy evaluation since the vignette builder crashes the first time it encounters a call to `GRanges`, which requires `GenomeInfoDb`, which in turn requires `GenomeInfoDbData`. I've tried fixing the problem by including `GenomeInfoDbData` in the package DESCRIPTION but this [also didn't help](http://bioconductor.org/spb_reports/openPrimeR_buildreport_20170824095655.html). The same thing happens on Travis-CI where the problem can be fixed by customizing the build procedure via manually requiring `GenomeInfoDbData`. A difference, however is, that Travis-CI throws the error sooner when checking whether the package dependencies are fulfilled following the installation and realizing that `GenomeInfoDbData` is not available. 

To sum up, there seems to be a problem with correctly determining/installing the `GenomeInfoDbData` dependency for `GenomicRanges` on OSX. Any pointers?

**Comment by vobencha** [comment]
Hi, 

Your package should not have to depend on, include or suggest the GenomeInfoDbData data package. openPrimeR would need to depend / import the GenomeInfoDb software package only if you directly call functions from GenomeInfoDb.

openPrimeR imports GenomicRanges in DESCRIPTION but does not fully or selectively import it in the NAMESAPCE. That needs to be fixed. If you're using functions and classes from GenomicRanges only as internal helpers then importing is fine. If a user of openPrimeR needs access to functions from GenomicRanges to operate on objects defined in openPrimeR then GenomicRanges should go in depends in DESCRIPTION.

I don't know why we'd see a difference between mac and linux wrt lazy loading in this case. I haven't looked at your package code in detail - please try fixing the imports first and we'll go from there.

Valerie

**Comment by matdoering** [comment]
Dear Valerie,
you're right about the GenomeInfoDb issue. The import in the NAMESPACE was just my failed attempt of trying to fix the OSX build error. Since then I've reverted back to the old DESCRIPTION with which Windows and Linux [pass all checks](http://bioconductor.org/spb_reports/openPrimeR_buildreport_20170824103707.html), while OSX still fails. Is that OK for now in terms of passing the checks?

Could you explain the first part of your second paragraph to me?

> openPrimeR imports GenomicRanges in DESCRIPTION but does not fully or selectively import it in the NAMESAPCE. That needs to be fixed. 

I only use GenomicRanges with the double-colon operator (i.e. GenomicRanges::), which is why I never added any of its exported symbols to the NAMESPACE. And no, I don't use functions from GenomicRanges to access any objects defined in the package.

When I tested building the pkg with MacOS the error occurred [here](https://github.com/matdoering/openPrimeR/blob/6733bc0a05f7fb258ddbecff4166e35ad10f57bb/R/con_primer_coverage.R#L1409]), the first time when calling the `GRanges` constructor in the `foreach` loop, complaining about `GenomeInfoDbData` not being installed.

**Comment by vobencha** [comment]
Historically we encouraged the use of selective imports in the NAMESPACE instead of the double colon operator, e.g., 
```
importFrom(GenomicRanges, GRanges)
```
Primary reasons were the accepted style at the time and clarity for what's being imported. With selective imports you can look at the NAMESPACE and quickly see what symbols the package was using. The current R documentation (https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Specifying-imports-and-exports) seems to equate the two approaches. The 'imports' field in DESCRIPTION makes sure the package is available during load time so in theory all exported symbols will be available. 

I won't push on this front - if you prefer to use the double colon operator that's fine. However I'd encourage you to be consistent. For example, Biostrings is in 'imports' in DESCRIPTION and one function (DNAStringSet) is selectively imported in the NAMESPACE. Then throughout the code you use '::' to call out many other functions from Biocstrings:

https://github.com/matdoering/openPrimeR/search?utf8=%E2%9C%93&q=Biostrings&type=

Either (1) fully import Biostrings in NAMESPACE or (2) selectively import from Biostrings in NAMESPACE or (3) don't import anything in NAMESPACE and use '::' for all functions from Biostrings.

I think the GenomeInfoDb problem may be related to R scoping rules in parallel evaluation. I would suggest printing out the environment inside the loop to see what packages/symbols are available. Related to parallel evaluation, it looks like in zzz.R you are circumventing parallel evaluation when on Windows. doParallel does support Windows ... was there a particular reason why you included this caveat?

I'll try testing on our Mac builder tomorrow.
Valerie

**Comment by matdoering** [comment]
Thanks, I'll make sure that the imports are more consistent in the next version. 

The GenomeInfoDb issue could be caused indeed be caused by the parallel call. Maybe the issue can already be solved by adding an importFrom(GenomicRanges, GRanges) to the NAMESPACE? I will test that in the next verison. I'd really appreciate it if you could look into it since I'm again without a Mac for testing.

The reason that parallel evaluations on Windows are disabled is that it it didn't work. Since it has been a couple of months when I encountered the issues, I cannot remember the exact error that was encountered  but I can clearly recall that some evaluations just didn't work at all when the parallel backend was used. I can even remember that I tried different parallel implementations but encountered problems with all of them, which is why I disabled parallel computations on Windows by default.

I've worked on creating a separate package for the Shiny frontend, which can be found [here]([https://github.com/matdoering/openPrimeRui). If I understand the Bioconductor documentation correctly, I can simply add this package here via `AdditionalPackage` once the initial package passes all checks, right?

Best
 Matthias

**Comment by matdoering** [comment]
I've just uploaded a new version that doesn't contain the Shiny app. I tried fixing the MacOS problem by loading GRanges outside the foreach loop; this didn't help in fixing the problem though. Any help on this would be highly appreciated.

**Comment by vobencha** [comment]
For testing on the mac I first stangled the vignette

```

veracruz1:vignettes biocbuild$ R CMD Stangle openPrimeR_vignette.Rmd
Output file:  openPrimeR_vignette.R

```

Then sourced the vignette and got this error:

```

> source("openPrimeR
openPrimeR/            openPrimeR_vignette.R 
> source("openPrimeR_vignette.R")
There are missing/non-functioning external tools.
To use the full potential of openPrimeR, please make sure
that all of the listed tools are installed:
o MELTING (http://www.ebi.ac.uk/biomodels/tools/melting/)
o OligoArrayAux (http://unafold.rna.albany.edu/OligoArrayAux.php)
o MAFFT (http://mafft.cbrc.jp/alignment/software/)
The number of cores for was set to '2' by 'parallel_setup()'.

| Task             | Templates | Primers      | Input file format  |
|------------------|-----------|--------------|--------------------|
| Design primers   | &#10003;  |              | FASTA, CSV         |
| Analyze primers  | &#10003;  | &#10003;     | FASTA, CSV         |
| Compare primers  | &#10003;  | &#10003;     | FASTA, CSV         |
Reading settings file: C_Taq_PCR_high_stringency.xml
Due to missing external tools the following constraints were ignored:cross_dimerization,self_dimerization
Due to missing external tools the following coverage constraints were ignored:coverage_model

| Slot             | Getter/Setter |  Purpose |
|------------------|---------------|----------|
| `Input_Constraints` | `constraints()` | Desired values for primer properties |
| `Input_Constraint_Boundaries` | `constraintLimits()`  | Limits for relaxing constraints during primer design |
| `Coverage_Constraints` | `cvg_constraints()` | Constraints for estimating the coverage |
| `PCR_conditions` | `PCR()`  | Experimental PCR conditions |
| `constraint_settings` | `conOptions()` | Settings for evaluating the constraints |

# Phase 1: Initialization of primers
# Phase 2: Selection of primers
a) Filtering of primers
    o primer_length: 49 primers remaining (100%)
    o gc_ratio: 37 primers remaining (75.51%)
    o no_runs: 37 primers remaining (75.51%)
    o no_repeats: 37 primers remaining (75.51%)
    o melting_temp_range: 37 primers remaining (75.51%)
Error in out$Structure_fw[fw.idx] <- str.fw$Structure :
  replacement has length zero

```

The error is coming from this line in the vignette:

```

## ----design_primers, message = FALSE, warning = FALSE, eval = TRUE-------
# Design forward primers for the first two templates
optimal.primers <- design_primers(template.df[1:2,], mode.directionality = "fw",
                                  settings = design.settings)

```

Jump into the offending function with the debugger:

```

> debug(openPrimeR:::compute.secondary.structures)

```

and we see this error:

```

Error in read.secondary.structure.raw(out.file) :
  The input file: /var/folders/gs/3s3km1v97wg7qy7rnhs782680000gp/T//RtmpEG0y8F/viennaRNA_fw_d42e5188d82a04007b0924d4771abae3109da4d6c63d7.txt.fold does not exist.

```

The problem is that 'result' in this chunk of code is always NULL because 'out.file' does not exist:

```

    ...

    out.file <- paste0(out.file, ".fold")
        if (file.exists(out.file)) {
                result <- read.secondary.structure.raw(out.file)
        } else {
                result <- NULL
        }
    return(result)
}


```

Debugging in read.secondary.structure.raw I see that you expect RNAfold to append the word ".fold" to the output file. I am not familiar with ViennaRNA. Is it a guarantee that RNAfold always appends ".fold" ... and evidently it write this file directly to disk? It looks like you can pass an output file name to RNAfold which seems more robust than assuming the default will always remain the same. It looks like you've taken some precautions but please make sure you are (always) using tempfile() and tempdir() instead of writing directly to disk.

Valerie

**Comment by vobencha** [comment]
Are you ready for me to look at this again?

Just fyi, if you're going to have parallel evaluation it should be supported across all platforms. There are several packages that support this. If you're getting a windows-specific error it would be good to see what that is.

Previously you asked if you should submit the shiny package to this issue. Yes, that's fine.

Valerie

**Comment by matdoering** [comment]
Dear Valerie,
yes, it would be great if you could take another look now. Here's a short summary of the changes so far:

* Modification of imports: If any method of a package is imported, any other required method is also used via *importFrom* rather than the double-colon operator; the only exception are functions with identical names from two packages 
* The ViennaRNA issue was related to the version of RNAfold installed on the Bioconductor servers, which was more recent than the one we were using, which caused problems due to different filename sanitizing under MacOS and modified write-out naming conventions (no more *.fold* appendix)
* The Shiny application was separated from the backend functionality. The frontend package is called *openPrimeRui* and will be added as an additional package in this issue.

Regarding parallelization under Windows, I will investigate the issues once again and get back to you with more details. 

Regarding the openPrimeRui package I was wondering whether there's any possibility of indicating that this is a frontend package such that BiocCheck warnings regarding runnable examples and evaluated vignette chunks are disabled? Since the package exports only a single function starting the shiny app, it's not possible to run this function or test it. To prevent the BiocCheck error due to too few runnable examples, I've added an example where the app is only run in interactive mode. The WARNING for the vignette still persists though.

Regarding the DESCRIPTION for the Shiny app - should it include the same SystemRequirements tag as the openPrimeR package or is it better to avoid the duplicated annotation?

Please note that I'm currently attaching the shinyBS package in UI.R because otherwise some shinyBS features do not work (e.g. bsButton, bsModal). For me it is not clear, why the functionality only works when shinyBS is attached, but I would like to prevent attaching the package if that is possible in any way.

I've looked at some Bioconductor repositories providing Shiny apps and I saw that most of them provide the code in the R/ folder itself rather than in inst/shiny. It seems that one advantage is that the NAMESPACE of the package is respected in this case, is that right? However, I'm not sure if it's possible to have a Docker container that directly starts the app independent of the package as then the NAMESPACE of the package won't be respected ... To summarize, I'm unsure what the most appropriate way of providing the code is. 

Thanks and best regards
 Matthias

**Comment by matdoering** [comment]
AdditionalPackage: https://github.com/matdoering/openPrimeRui

**Comment by matdoering** [comment]
Two small updates:

* Parallelization is now also active under Windows. I initially deactivated it under Windows because there was an unserialize error during the parallel SOCK cluster computations of the coverage. It turned out that the problem was caused by the *updateProgress* callback function, which was active in the foreach loop. After removing the callback in the parallel call, the problem was gone. This was probably due to the parallel foreach backend being different for Windows than for Linux, where the problem doesn't occur.
* It seems that it's necessary to attach shinyBS for some functionalities (e.g. using updateProgress); A minimal working example can be found [here](https://github.com/ebailey78/shinyBS/issues/92).

**Comment by vobencha** [comment]
Sorry for the delay. I'll take a look at this today.
Valerie

**Comment by vobencha** [comment]
Hi Matthias,

Thank you for making the changes, fixing parallel evaluation on Windows, etc. Did you add a webhook to openPrimeRui? I ask because we didn't get a build for it. If you think you've got it sorted I'll ask Lori to kick off another build.

- openPrimeRui only needs to Depend on openPrimeR (as you've done). You don't need to re-list the system requirements.

- I would suggest not using LazyData: true in DESCRIPTION
This causes large packages to load very, very slowly. I've never seen this option to be an advantage - only one to cause problems.

- If possible, add runable examples to these flagged man pages.
```
* Checking exported objects have runnable examples...
    * NOTE: Consider adding runnable examples to the following man
      pages which document exported objects:
      parallel_setup.Rd, plot_conservation.Rd, runTutorial.Rd,
  score_conservation.Rd, score_degen.Rd,
  select_regions_by_conservation.Rd
```
If they must be interactive you can wrap them in
```
if (interactive()) {
...
}
```
- Please consolidate the man pages by topic. There are currently 405 pages. Let's aim for somewhere less than 20.
```
~/repos/git/software/openPrimeR/man >ls|wc
    405     405    9684
```
- As you mentioned, you'll need a vignette for the openPrimeRui package. It can be a static These are a couple of examples that may help

    https://www.bioconductor.org/packages/devel/bioc/vignettes/cytofkit/inst/doc/cytofkit_shinyAPP.html
    http://bioconductor.org/packages/release/bioc/vignettes/twoddpcr/inst/doc/twoddpcr.html

    Also, maybe useful is Andrzej's tutorial here:

    https://www.bioconductor.org/help/course-materials/2015/CSAMA2015/lab/shiny.html#share-as-r-files

- In response to this question:

"I've looked at some Bioconductor repositories providing Shiny apps and I saw that most of them provide the code in the R/ folder itself rather than in inst/shiny. It seems that one advantage is that the NAMESPACE of the package is respected in this case, is that right? However, I'm not sure if it's possible to have a Docker container that directly starts the app independent of the package as then the NAMESPACE of the package won't be respected ... "

I would avoid the use of the Docker. We can't support that on the build system.

I'm going to ask Andrzej's opinion here. He has quite a bit of experience with Shiny.

@aoles, have you worked with shinyBS? And if you have time, any comments on the structure of openPrimeRui wrt the location of server.R, ur.R etc.?

Valerie

**Comment by matdoering** [comment]
AdditionalPackage: https://github.com/matdoering/openPrimeRui

**Comment by matdoering** [comment]
Dear Valerie,
regarding the openPrimeRui build I was already wondering why the build didn't start when I added it with *AdditionalPackage* initially. At that time, I didn't have the webhook set yet and when I set it later, it started transmitting data but didn't initialize the build. Hence, I just added another comment with *AdditionalPackage* to initialize the build and that made it work. 

According to your suggestion, I just removed *Lazydata: true* for both packages and added runnable examples to the functions with missing examples - using *if interactive()* is extremely helpful. 

The reason that there were 405 man/ files in the package was that these files included docs for many internal functions. Since the package itself doesn't expose these functions, I simply removed the corresponding .Rd files from the man/ folder. With this change, there's only 81 Rd files remaining. Is that sufficient? You mentioned that I should aim at less than 20, but I don't see how merging that many documentation files would be possible. 

I've already created a vignette for the *openPrimeRui* package. Right now there's a warning because too few chunks are evaluated. In the new version, I've fixed this by requiring an interactive session when evaluating the chunk. 

Thank you for asking about insights on the recommended package structure for Shiny apps and the shinyBS issue. I was just wondering about Docker because I was wondering if there was any limitation on the package structure as to running the Shiny app in a container. With the current structure of the package, I couldn't use the package namespace for many things as all the shiny code is located in inst/exdata/shiny, which can be quite bothersome, although it works fine.

Best Matthias

**Comment by matdoering** [comment]
I was wondering whether there's any updates on the review for _openPrimeRui_ and whether the changes I've made to _openPrimeR_ are already satisfactory. Looking forward to your feedback!

**Comment by vobencha** [comment]
I've been out of the office. I'll try to look at this in the next couple of days.
Valerie

**Comment by vobencha** [comment]
Hi,
openPrimeR looks good except for the man pages. Thank you for removing the non-exported function pages -
 that's great. We are still left with a large number and I'm wondering why you feel combining them is a problem. Do you think it's too much info per page or maybe you don't think they fall into logical categories ...? My concern is how will the user find all of these functions. The vignette can't cover the functionality of 80 functions so having them grouped on a single man page (e.g., all plot functions together) makes discovery easier.

Other than that we are good to go.
Valerie

**Comment by matdoering** [comment]
Dear Valerie,
I just updated the documentation of the package such that the manual entries are now logically consolidated. The PDF manual now only contains a total of 13 entries including the package description and the index. Please find the new listing attached. Is it fine like that?

Are there any updates regarding `openPrimeRui`?


![doc_entries](https://user-images.githubusercontent.com/25709555/30804982-cf356912-a1ef-11e7-977a-6ca67e8088b8.png)

**Comment by vobencha** [comment]
Thanks for consolidating the man pages. No more updates for openPrimeRui. Please bump the versions on both packages so we get 2 clean builds. Once that's done we're good to go.
Valerie

**Comment by matdoering** [comment]
Dear Valerie,
great! I've just bumped the versions of both packages and they were built successfully.

Thank you so much for your guidance
 Matthias

**Comment by vobencha** [comment]
Great. All set. Thanks for your work on these packages.
Valerie

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/matdoering.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://docs.google.com/forms/d/e/1FAIpQLSdlTbNjsQJDp0BA480vo4tNufs0ziNyNmexegNZgNieIovbAA/viewform

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"openPrimeR\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/openPrimeR

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

**Comment by matdoering** [comment]
Dear Martin, 
first of thank you for adding openPrimeR to Bioconductor. Since I didn't receive any information about openPrimeRui being added to Bioconductor I was wondering whether it is usual for additional packages being added later than the 'main package' or whether it would be necessary to open another issue for openPrimeRui? Thanks for the clarification!

**Comment by mtmorgan** [comment]
Sorry, our workflow failed to catch the AdditionalPackage; I'll work on this today.

**Comment by mtmorgan** [comment]
The problem was that the workflow only looked for additional packages in the first 30 comments; openPrimeR has been added to the Bioconductor git repository and manifest for nightly builds; it should be included in Thursday's build report.
